### PR TITLE
Delegated Credentials (WORK IN PROGRESS)

### PR DESCRIPTION
**Pull request by tsusanka**
This is not yet finished, this PR is more of a "checkpoint".

TODO: 
- [x] saving the credentials to use them until they're valid
- [ ] test edge cases
- [ ] TLS 1.3 support
- [ ] incorporate into keyless

cc @grittygrease

### CODE REVIEW

File: handshake_server.go [hunk 1]
```diff
@@ -356,6 +369,24 @@ Curves:
 		}
 	}
 
+	if hs.clientHello.delegatedCredentials {
```
Comment (grittygrease) [inline handshake_server.go]: This logic should most likely be moved fully into GetCertificate. A cleaner design would be to simply create a new PrivateKey in the certificate object that corresponds to the delegated credential private key. That way almost all logic on the server side would be unchanged, aside from the addition of a new opaque delegated credentials object in the serverhello.

File: handshake_delegated_cred.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"time"
+)
+
```
Comment (grittygrease) [inline handshake_delegated_cred.go]: This tight integration with the handshake is not ideal. I think all the credential minting and verifying should be abstracted into its own package (byte slice + cert to PrivateKey and back). The only pieces that need to change on the client side are the sending of the empty extension, the extraction and validation of the delegated credential, and the CertificateVerify function.

File: common.go [hunk 3]
```diff
@@ -11,6 +11,7 @@ import (
 	"crypto/rand"
 	"crypto/sha512"
 	"crypto/x509"
+	"encoding/asn1"
```
Comment (grittygrease) [inline common.go]: As mentioned below, this is not a great dependency to add. Factor out as much logic as possible into its own package and try to keep the changes to crypto/tls minimal.

File: common.go [hunk 4]
```diff
@@ -949,6 +954,9 @@ type Certificate struct {
 	// OCSPStaple contains an optional OCSP response which will be served
 	// to clients that request it.
 	OCSPStaple []byte
+	// DelegatedCredential contains a signed object containing a public
+	// key for signing handshakes
+	DelegatedCredential DelegatedCredential
```
Comment (grittygrease) [inline common.go]: This is probably the wrong place for this type of structure. When the Certificate is returned from GetCertificate it should already have a serialized opaque delegated credential, not a structure that needs to be serialized.

File: handshake_delegated_cred_test.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Comment (grittygrease) [inline handshake_delegated_cred_test.go]: Again, this is too tightly integrated with the handshake. Each individual function should be unit tested with respect to its functionality, the internal tls structures should not be involved.

**Comment by grittygrease** [comment]
Config.Clone and other ancillary functions are also missing the new data.

**Comment by grittygrease** [comment]
Also note that the tests are failing.

File: common.go [hunk 6]
```diff
@@ -614,6 +616,9 @@ type Config struct {
 	// this Config is returned by a GetConfigForClient callback. It's used
 	// by serverInit in order to copy session ticket keys if needed.
 	originalConfig *Config
+
+	// states if Delegated Credentials should be used
+	useDelegatedCredentials bool
```
Comment (tsusanka) [inline common.go]: You're right, however, how come I still can access it? https://github.com/tsusanka/tls-tris/blob/tsusanka/delegated-cred/handshake_delegated_cred_test.go#L254

File: common.go [hunk 7]
```diff
@@ -959,6 +967,23 @@ type Certificate struct {
...
 }
 
+type DelegatedCredential struct {
+	ValidTime int64 // todo - change to Time?
+	PublicKey interface{}
+}
+
+// checks certificate if it contains the DelegationUsage extension
+// required for Delegated Credentials
+func (hs *serverHandshakeState) isCertificateValidForDelegationUsage() bool {
```
Comment (tsusanka) [inline common.go]: Got it. The content of the handshake_delegated_cred.go file will be then moved to separate crypto/tls/delcred (or similar name) sub package. This sub package will be responsible for creating, marshalling, signing, validating etc. of the credentials.

File: handshake_delegated_cred_test.go (hunk 5 above)
Comment (tsusanka) [inline handshake_delegated_cred_test.go]: Okay. So I should add unit tests - e.g. check if the credential marshalled correctly etc. What is wrong with those higher-level tests though? They still provide value, don't they?

File: common.go (hunk 4 above)
Comment (tsusanka) [inline common.go]: I understand. The Certificate struct will then contain marshalled DC and the associated PrivateKey. Both received from the delcred sub package.

File: handshake_server.go (hunk 1 above)
Comment (tsusanka) [inline handshake_server.go]: Thanks. I'm a little confused, how to use the GetCertificate function. If the c.GetCertificate is not null it is called to retrieve the certificate.

So, my goal is to implement a new GetCertificate function which I then "override" how the certificate is retrieved, right? The function will add the delegated credential and its PrivateKey. However, where this code should be? Also in the subpackage? Where do I then set the config to use this function?

File: handshake_delegated_cred.go (hunk 2 above)
Comment (tsusanka) [inline handshake_delegated_cred.go]: >  The only pieces that need to change on the client side are the sending of the empty extension, the extraction and validation of the delegated credential, and the CertificateVerify function.

I'm confused about the CertificateVerify part. Do you mean the CertificateVerify message? But that's used for client authentication, isn't it?

**Comment by tsusanka** [comment]
Thanks for the review. I'm aware of the failing tests, it is some kind of race condition issue, I'll have a look into that once I fix the outlined issues.

File: handshake_server.go (hunk 1 above)
Comment (tsusanka) [inline handshake_server.go]: Ah, alright, I simply set the function in the config when spinning up TLS, right?

File: common.go (hunk 4 above)
Comment (grittygrease) [inline common.go]: Perhaps consider reusing the existing PrivateKey struct.

File: common.go (hunk 7 above)
Comment (grittygrease) [inline common.go]: Yes, that sounds like a plan.

File: handshake_delegated_cred.go (hunk 2 above)
Comment (grittygrease) [inline handshake_delegated_cred.go]: Actually, in TLS 1.2 this is called the ServerKeyExchange.

File: cipher_suites.go [hunk 8]
```diff
@@ -33,7 +34,7 @@ type keyAgreement interface {
 
 	// This method may not be called if the server doesn't send a
 	// ServerKeyExchange message.
-	processServerKeyExchange(*Config, *clientHelloMsg, *serverHelloMsg, *x509.Certificate, *serverKeyExchangeMsg) error
+	processServerKeyExchange(*Config, *clientHelloMsg, *serverHelloMsg, crypto.PublicKey, *serverKeyExchangeMsg) error
```
Comment (tsusanka) [inline cipher_suites.go]: It might be possible to avoid this change by wrapping the DC's PublicKey into an x509.Certificate. However:
a) I'm not sure it'll work since we have only the PublicKey
b) all processServerKeyExchange functions use only the public key, so the dependency for x509.Certificate seems  redundant
c) the function is not exported, so it should not be a BC break (?)

**Comment by tsusanka** [comment]
I think this is ready for another round of code review.
- All code is in the same _tls_ package, however, the logic is clearly separated using the GetCertificate function as discussed
- I've squashed all commits into just a few, it was a mess
- Tests are fixed
- I've implemented the support for PKCS1, it's trivial to remove it in case we decide to

File: cipher_suites.go (hunk 8 above)
Comment (grittygrease) [inline cipher_suites.go]: This is a slick change. I can't see why it wouldn't work or why *x509.Certificate was used in the first place.

File: delegated_credentials.go [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+type GetCertificate func(*ClientHelloInfo) (*Certificate, error)
+
+type DelegatedCredential struct {
+	ValidTime int64
+	PublicKey interface{}
+}
+
+func NewDelCredGetCertificateFunction(cert *Certificate) GetCertificate {
+
```
Comment (grittygrease) [inline delegated_credentials.go]: Unnecessary space?

File: delegated_credentials.go [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+var DelegatedCredentialsIdentifier = asn1.ObjectIdentifier{2, 5, 29, 99}
+
+type GetCertificate func(*ClientHelloInfo) (*Certificate, error)
+
+type DelegatedCredential struct {
+	ValidTime int64
+	PublicKey interface{}
+}
+
+func NewDelCredGetCertificateFunction(cert *Certificate) GetCertificate {
```
Comment (grittygrease) [inline delegated_credentials.go]: This abbreviation is a bit awkward. NewGetCertificateFromDelegatedCredential? That's also long but a bit better.

File: delegated_credentials.go [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+	return func(clientHelloInfo *ClientHelloInfo) (*Certificate, error) {
+
```
Comment (grittygrease) [inline delegated_credentials.go]: Space

File: delegated_credentials.go [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+		}
+
+		cert.DelegatedCredential = credentialBytes
+		cert.PrivateKey = privateKey
+		return cert, nil
+	}
+}
+
+func selectVersion(versions []uint16) uint16 {
+
```
Comment (grittygrease) [inline delegated_credentials.go]: Newline

File: delegated_credentials.go [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+	validTill := time.Now().Add(CredentialsValidity)
+	relativeToCert := validTill.Sub(certificate.NotBefore)
+	credential = DelegatedCredential{
+		ValidTime: int64(relativeToCert.Seconds()),
+	}
+
+	if scheme == ECDSAWithP256AndSHA256 {
+		privateKey, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		credential.PublicKey = privateKey.(crypto.Signer).Public()
+
```
Comment (grittygrease) [inline delegated_credentials.go]: newline

File: delegated_credentials.go [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+	publicKeyBytes, err := x509.MarshalPKIXPublicKey(dc.PublicKey)
+	publicKeyLength := len(publicKeyBytes)
+	cred[4] = uint8(publicKeyLength >> 16)
+	cred[5] = uint8(publicKeyLength >> 8)
+	cred[6] = uint8(publicKeyLength)
+
+	copy(cred[7:7+publicKeyLength], publicKeyBytes)
+
+	cred[publicKeyLength+7] = uint8(scheme >> 8) // hash
```
Comment (grittygrease) [inline delegated_credentials.go]: This will likely be different for TLS 1.3

File: handshake_client.go [hunk 15]
```diff
@@ -21,13 +21,14 @@ import (
...
-	masterSecret []byte
-	session      *ClientSessionState
+	c                *Conn
+	serverHello      *serverHelloMsg
+	hello            *clientHelloMsg
+	suite            *cipherSuite
+	finishedHash     finishedHash
+	masterSecret     []byte
+	session          *ClientSessionState
+	delCredPublicKey crypto.PublicKey
```
Comment (grittygrease) [inline handshake_client.go]: Maybe this should be just connection PublicKey, and it is set to either the delegated cred key or c.peerCertificates[0].PublicKey. Then you may be able to reduce the number of if statements below.

File: delegated_credentials.go (hunk 10 above)
Comment (tsusanka) [inline delegated_credentials.go]: Yeah, I don't really like it myself. I'm not sure the `From` is accurate. So how about simply `NewDelegatedCredentialGetCertificate`?

File: delegated_credentials.go [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"time"
+)
+
+// After only renewThreshold of credential's validTime is left a new one is created
+const renewThreshold = 0.2
```
Comment (tsusanka) [inline delegated_credentials.go]: Maybe better to reverse this? Set it to `0.8` and call it `renewAfterThreshold`

File: delegated_credentials.go [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+		}
+	}
+	return 0
+}
+
+// Creates new Delegated Credential. The type of the credential is decided based on the selected
+// SignatureScheme to ensure client's support.
+func createDelegatedCredential(config *CredentialsConfig) (credential DelegatedCredential, privateKey crypto.PrivateKey, renewAfter time.Time, err error) {
+	validTill := config.timeFunc().Add(config.Validity)
+	relativeToCert := validTill.Sub(config.Cert.Leaf.NotBefore)
```
Comment (grittygrease) [inline delegated_credentials.go]: The computation of the ValidTime is a rather complex dance. Is there a simpler way to compute it that is more readable?

File: delegated_credentials.go [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+type CredentialsConfig struct {
+	Cert      *Certificate
+	Validity  time.Duration
+	scheme    SignatureScheme
+	version   uint16
+	timeFunc  func() time.Time
+	renewChan chan struct{}
+}
+
+var cachedCredentials map[SignatureScheme]*cachedCredential
```
Comment (grittygrease) [inline delegated_credentials.go]: This is a package-global variable indexed by the SignatureScheme. This means it will be shared between all instantiations, even those with different certs. You most likely want to keep this local to the CredentialsConfig.

File: delegated_credentials.go [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+		cachedCredentials = make(map[SignatureScheme]*cachedCredential)
+	}
+	if cachedCredentials[config.scheme] == nil {
+		cachedCredentials[config.scheme], err = newCredential(config)
+	} else {
+		if config.timeFunc().After(cachedCredentials[config.scheme].renewAfter) {
+			go func() {
+				cachedCredentials[config.scheme], err = newCredential(config)
+				if config.renewChan != nil {
+					config.renewChan <- struct{}{}
```
Comment (grittygrease) [inline delegated_credentials.go]: Where is the renewChan read?

File: delegated_credentials.go (hunk 18 above)
Comment (tsusanka) [inline delegated_credentials.go]: Good point! So I'm just including it as a not exported CredentialsConfig variable.

File: delegated_credentials.go (hunk 19 above)
Comment (tsusanka) [inline delegated_credentials.go]: Only in [tests](https://github.com/cloudflare/tls-tris/pull/29/files#diff-74969c972e0630d3c7b131f8f82325e9R270). I was reluctant to add the channel just because of tests, but I wanted to test that the new credential is created, which I can do only after the goroutine finishes. I was talking to Salmen about this and he advised me this.

File: delegated_credentials.go (hunk 17 above)
Comment (tsusanka) [inline delegated_credentials.go]: Yeah, I agree. I was thinking about it and I'm not sure there is a better way though. The code is basically the definition in itself:
> Relative time in seconds from the beginning of the certificate's notBefore value after which the Delegated Credential is no longer valid

Btw why is this relative? Why not making it a simple datetime when the credential expires? The client needs to verify the certificate's validity anyway.

----

I can remove the `relativeToCert` variable but it's not significantly better:

```go
validTill := config.timeFunc().Add(config.Validity)
renewAfter = validTill.Add(-time.Duration(config.Validity.Seconds()*renewThreshold) * time.Second)
credential = DelegatedCredential{
	ValidTime: int64(validTill.Sub(config.Cert.Leaf.NotBefore).Seconds()),
}
```

**Comment by tsusanka** [comment]
Closed in favour #32
